### PR TITLE
feat: split users into human and machine

### PR DESCRIPTION
**Pull request by adlerhurst**

### CODE REVIEW

File: pkg/grpc/management/proto/management.proto [hunk 1]
```diff
@@ -1473,8 +1637,8 @@ message UserProfileView {
   uint64 sequence = 9;
   google.protobuf.Timestamp creation_date = 10;
   google.protobuf.Timestamp change_date = 11;
-  repeated string login_names = 12;
-  string preferred_login_name = 27;
+  repeated string login_names = 12;//
+  string preferred_login_name = 27;//
```
Comment (fgerschwiler) [inline pkg/grpc/management/proto/management.proto:1641]: //?

**Review by fgerschwiler — COMMENTED** [review]
add tests, in my opinion admin part should be finished with this pr, auth is possible on other pr, in this case keys and views should be included here

File: migrations/cockroach/V1.99__user_machine.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+ALTER TABLE management.users ADD COLUMN machine_name STRING, ADD COLUMN machine_description STRING, ADD COLUMN user_type STRING;
+ALTER TABLE adminapi.users ADD COLUMN machine_name STRING, ADD COLUMN machine_description STRING, ADD COLUMN user_type STRING;
+ALTER TABLE auth.users ADD COLUMN machine_name STRING, ADD COLUMN machine_description STRING, ADD COLUMN user_type STRING;
+
+--TODO: (adminapi)iam_members
+--TODO: (auth,authz,management)user_grants
+--TODO: (auth)user_sessions -> maybe no changes needed
+--TODO: (management)project_grant_members 
+--TODO: (management)org_members
+--TODO: (management)project_members
```
Comment (livio-a) [inline migrations/cockroach/V1.99__user_machine.sql:10]: TODO

File: pkg/grpc/management/proto/management.proto [hunk 3]
```diff
@@ -71,6 +71,37 @@ service ManagementService {
...
+
+// rpc DeleteServiceAccountKey(ServiceAccountKeyIDRequest) returns (ServiceAccountKeyResponse) {
+//     option (google.api.http) = {
+//         post: "/serviceaccounts/{service_account_id}/keys/{key_id}"
+//     };
+
+//     option (caos.zitadel.utils.v1.auth_option) = {
+//         permission: "serviceaccount.write"
+//     };
+// }
```
Comment (livio-a) [inline pkg/grpc/management/proto/management.proto:93]: ?

File: pkg/grpc/management/proto/management.proto [hunk 4]
```diff
@@ -1375,40 +1451,127 @@ message UserView {
...
+  string description = 4;
+  repeated MachineKeyView keys = 5;
+}
+
+message UpdateMachineRequest {
+  string id = 1;
+  string description = 2 [(validate.rules).string.max_len = 500];
+}
+
+// full key
```
Comment (livio-a) [inline pkg/grpc/management/proto/management.proto:1524]: ?

File: pkg/grpc/management/proto/management.proto [hunk 5]
```diff
@@ -1567,6 +1731,35 @@ message UpdateUserAddressRequest {
...
+// message ServiceAccountKeyResponse {
+//   string key_id = 1;
+//   string service_account_id = 2;
+//   google.protobuf.Timestamp expiration_date = 3;
+
+//   google.protobuf.Timestamp creation_date = 4;
+//   google.protobuf.Timestamp change_date = 5;
+//   uint64 sequence = 6;
+//   string resource_owner = 7;
+// }
```
Comment (livio-a) [inline pkg/grpc/management/proto/management.proto:1745]: ?

**Review by livio-a — CHANGES_REQUESTED** [review]
same as Fabi... add some specific tests for human and machine (as wrong input)

File: internal/auth/repository/eventsourcing/handler/user.go [hunk 6]
```diff
@@ -56,7 +56,9 @@ func (u *User) ProcessUser(event *models.Event) (err error) {
 	user := new(view_model.UserView)
 	switch event.Type {
 	case es_model.UserAdded,
-		es_model.UserRegistered:
+		es_model.UserRegistered,
+		es_model.MachineAdded,
+		es_model.HumanAdded:
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/auth/repository/eventsourcing/handler/user.go:62]: done

File: internal/auth/repository/eventsourcing/handler/user.go [hunk 7]
```diff
@@ -77,7 +79,19 @@ func (u *User) ProcessUser(event *models.Event) (err error) {
...
+		es_model.HumanEmailChanged,
+		es_model.HumanEmailVerified,
+		es_model.HumanPhoneChanged,
+		es_model.HumanPhoneVerified,
+		es_model.HumanPhoneRemoved,
+		es_model.HumanAddressChanged,
+		es_model.HumanMfaOtpAdded,
+		es_model.HumanMfaOtpVerified,
+		es_model.HumanMfaOtpRemoved,
+		es_model.MachineChanged,
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/auth/repository/eventsourcing/handler/user.go:94]: done

File: internal/management/repository/eventsourcing/eventstore/user.go [hunk 8]
```diff
@@ -163,9 +165,16 @@ func (repo *UserRepo) ProfileByID(ctx context.Context, userID string) (*usr_mode
 	if err != nil {
 		return nil, err
 	}
+	if user.HumanView == nil {
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/eventstore/user.go:168]: handeled in get<subobject>

File: internal/management/repository/eventsourcing/handler/user.go [hunk 9]
```diff
@@ -55,7 +55,9 @@ func (u *User) ProcessUser(event *models.Event) (err error) {
 	user := new(view_model.UserView)
 	switch event.Type {
 	case es_model.UserAdded,
-		es_model.UserRegistered:
+		es_model.UserRegistered,
+		es_model.MachineAdded,
+		es_model.HumanAdded:
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/user.go:61]: done

File: pkg/grpc/management/proto/management.proto [hunk 10]
```diff
@@ -1495,7 +1659,7 @@ message UserEmail {
   google.protobuf.Timestamp change_date = 6;
 }
 
-message UserEmailView {
+message UserEmailView { //equal
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply pkg/grpc/management/proto/management.proto]: removed

File: pkg/grpc/management/proto/management.proto (hunk 1 above)
Comment (adlerhurst) [reply pkg/grpc/management/proto/management.proto]: removed

File: internal/api/grpc/management/user_machine_converter.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+	expirationDate, err := ptypes.TimestampProto(key.CreationDate)
+	logging.Log("MANAG-gluk7").OnError(err).Debug("unable to parse timestamp")
+
+	return &management.MachineKeyView{
+		Id:             key.ID,
+		CreationDate:   creationDate,
+		ExpirationDate: expirationDate,
+		Sequence:       key.Sequence,
+		Type:           management.MachineKeyType_MACHINEKEY_UNSPECIFIED, //TODO: type mapping
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/api/grpc/management/user_machine_converter.go]: done

File: internal/management/repository/eventsourcing/handler/project_grant_member.go [hunk 12]
```diff
@@ -121,6 +124,7 @@ func (p *ProjectGrantMember) fillData(member *view_model.ProjectGrantMemberView)
 	return nil
 }
 
+//TODO: specific for user data
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/project_grant_member.go]: removed

File: internal/management/repository/eventsourcing/handler/project_member.go [hunk 13]
```diff
@@ -121,6 +124,7 @@ func (p *ProjectMember) fillData(member *view_model.ProjectMemberView) (err erro
 	return nil
 }
 
+//TODO: secific for user data
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/project_member.go]: removed

File: internal/management/repository/eventsourcing/handler/user_grant.go [hunk 14]
```diff
@@ -161,6 +164,7 @@ func (u *UserGrant) fillData(grant *view_model.UserGrantView, resourceOwner stri
 	return nil
 }
 
+//TODO: secific for human or machine
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/user_grant.go]: removed

File: internal/user/repository/eventsourcing/eventstore.go [hunk 15]
```diff
@@ -660,6 +751,9 @@ func (es *UserEventstore) ChangeEmail(ctx context.Context, email *usr_model.Emai
 	if err != nil {
 		return nil, err
 	}
+	if existing.Human == nil {
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/eventsourcing/eventstore.go]: done

File: internal/user/repository/eventsourcing/eventstore.go [hunk 16]
```diff
@@ -694,6 +788,9 @@ func (es *UserEventstore) VerifyEmail(ctx context.Context, userID, verificationC
 	if err != nil {
 		return err
 	}
+	if existing.Human == nil {
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/eventsourcing/eventstore.go]: done

File: internal/user/repository/eventsourcing/eventstore.go [hunk 17]
```diff
@@ -726,6 +823,9 @@ func (es *UserEventstore) CreateEmailVerificationCode(ctx context.Context, userI
...
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/eventsourcing/eventstore.go]: done

File: internal/user/repository/eventsourcing/model/mfa.go [hunk 18]
```diff
@@ -32,18 +33,18 @@ func OTPToModel(otp *OTP) *model.OTP {
 	}
 }
 
-func (u *User) appendOtpAddedEvent(event *es_models.Event) error {
+func (u *Human) appendOtpAddedEvent(event *es_models.Event) error {
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/eventsourcing/model/mfa.go]: done

File: migrations/cockroach/V1.99__user_machine.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (adlerhurst) [reply migrations/cockroach/V1.99__user_machine.sql]: migration 1.10 is final edition

File: internal/admin/repository/eventsourcing/handler/iam_member.go [hunk 19]
```diff
@@ -116,6 +119,7 @@ func (m *IamMember) fillData(member *iam_model.IamMemberView) (err error) {
 	return nil
 }
 
+//TODO: specific for user data
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/admin/repository/eventsourcing/handler/iam_member.go]: removed

File: internal/api/grpc/management/user_machine_converter.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/api/grpc/management/user_machine_converter.go]: removed

File: internal/auth/repository/eventsourcing/handler/user_grant.go [hunk 20]
```diff
@@ -326,6 +329,7 @@ func (u *UserGrant) fillData(grant *view_model.UserGrantView, resourceOwner stri
 	return nil
 }
 
+//TODO: specific for user data
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/auth/repository/eventsourcing/handler/user_grant.go]: done

File: internal/management/repository/eventsourcing/handler/project_grant_member.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/project_grant_member.go]: done

File: internal/management/repository/eventsourcing/handler/project_member.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/project_member.go]: done

File: internal/management/repository/eventsourcing/handler/user_grant.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/management/repository/eventsourcing/handler/user_grant.go]: don

File: internal/user/repository/eventsourcing/model/user_machine.go [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+func (sa *Machine) AppendEvent(event *models.Event) (err error) {
+	switch event.Type {
+	case MachineAdded, MachineChanged:
+		err = sa.setData(event)
+	case KeyAdded:
+		fallthrough
+	case KeyRemoved:
+		logging.Log("MODEL-iBgOc").Panic("key unimplemented")
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/eventsourcing/model/user_machine.go]: implemented

File: internal/user/repository/view/model/user.go [hunk 22]
```diff
@@ -26,115 +27,118 @@ const (
...
+			MfaInitSkipped:         user.MfaInitSkipped,
+			InitRequired:           user.InitRequired,
+		}
 	}
+
+	if !user.MachineView.IsZero() {
+		userView.MachineView = &model.MachineView{
+			Description: user.MachineView.Description,
+			Name:        user.MachineView.Name,
+			//TODO: keys
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/repository/view/model/user.go]: done

File: migrations/cockroach/V1.99__user_machine.sql (hunk 2 above)
Comment (adlerhurst) [reply migrations/cockroach/V1.99__user_machine.sql]: migration 1.10 is final edition

File: pkg/grpc/management/proto/management.proto (hunk 3 above)
Comment (adlerhurst) [reply pkg/grpc/management/proto/management.proto]: done

File: pkg/grpc/management/proto/management.proto (hunk 4 above)
Comment (adlerhurst) [reply pkg/grpc/management/proto/management.proto]: implemented

File: pkg/grpc/management/proto/management.proto (hunk 5 above)
Comment (adlerhurst) [reply pkg/grpc/management/proto/management.proto]: implemented

**Review by fgerschwiler — APPROVED** [review]

**Review by livio-a — APPROVED** [review]

**Pull request merged**